### PR TITLE
Fixed BLAS and LAPACK configuration on FreeBSD

### LINKED ISSUE #7330: conf-blas & conf-lapack fail to install under FreeBSD
When installing `conf-blas` under FreeBSD 10 amd64 I get this error:

```
$ opam install conf-blas                                                                                            
The following actions will be performed:
  ∗  install conf-blas 1

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[ERROR] The compilation of conf-blas failed at "sh -exc cc $CFLAGS test.c -lblas".

#=== ERROR while installing conf-blas.1 =======================================#
# opam-version 1.2.2
# os           freebsd
# command      sh -exc cc $CFLAGS test.c -lblas
# path         ~/.opam/4.03.0+flambda/build/conf-blas.1
# compiler     4.03.0+flambda
# exit-code    1
# env-file     ~/.opam/4.03.0+flambda/build/conf-blas.1/conf-blas-32329-ccabea.env
# stdout-file  ~/.opam/4.03.0+flambda/build/conf-blas.1/conf-blas-32329-ccabea.out
# stderr-file  ~/.opam/4.03.0+flambda/build/conf-blas.1/conf-blas-32329-ccabea.err
### stderr ###
# + cc test.c -lblas
# /usr/bin/ld: cannot find -lblas
# cc: error: linker command failed with exit code 1 (use -v to see invocation)



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions failed
  ∗  install conf-blas 1
No changes have been performed

=-=- conf-blas.1 troobleshooting =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
=> This package relies on external (system) dependencies that may be missing. `opam depext conf-blas.1' may help you find the correct
   installation for your system.
```

Adding `-L/usr/local/lib` to `CFLAGS`, so the build script would find `libblas.so`, results in a new set of errors:

```
#=== ERROR while installing conf-blas.1 =======================================#
# opam-version 1.2.2
# os           freebsd
# command      sh -exc cc $CFLAGS test.c -lblas
# path         /home/magv/.opam/4.03.0+flambda/build/conf-blas.1
# compiler     4.03.0+flambda
# exit-code    1
# env-file     ~/.opam/4.03.0+flambda/build/conf-blas.1/conf-blas-32343-ccabea.env
# stdout-file  ~/.opam/4.03.0+flambda/build/conf-blas.1/conf-blas-32343-ccabea.out
# stderr-file  ~/.opam/4.03.0+flambda/build/conf-blas.1/conf-blas-32343-ccabea.err
### stderr ###
# [...]
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__addtf3@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__gttf2@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__divtf3@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__letf2@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__netf2@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__floatditf@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__trunctfdf2@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__eqtf2@GCC_4.6.0'
# //usr/local/lib/gcc48/libgfortran.so.3: undefined reference to `__floatsitf@GCC_4.6.0'
# cc: error: linker command failed with exit code 1 (use -v to see invocation)
```

This is probably because BLAS under FreeBSD is build with GCC, while the default `cc` is CLANG, and there seem to be incompatibilities between them. Whatever the case, here's a change to `conf-blas` that seems to make it work for me:

``` diff
--- ~/.opam/repo/default/packages/conf-blas/conf-blas.1/opam.orig      2016-08-31 15:26:25.000000000 +0300
+++ ~/.opam/repo/default/packages/conf-blas/conf-blas.1/opam   2016-08-31 15:42:09.000000000 +0300
@@ -6,7 +6,8 @@
 dev-repo: "https://github.com/ocaml/opam-repository.git"
 license: "GPL-2.0"
 build: [
-  ["sh" "-exc" "cc $CFLAGS test.c -lblas"] {os != "darwin" & os != "win32"}
+  ["sh" "-exc" "cc $CFLAGS test.c -lblas"] {os != "darwin" & os != "freebsd" & os != "win32"}
+  ["sh" "-exc" "${CC:-gcc} $CFLAGS test.c -lblas $LDFLAGS"] {os = "freebsd"}
   ["sh" "-exc" "cc -framework Accelerate $CFLAGS test.c -lblas"] {os = "darwin"}
   ["%{build}%/test-win.sh"] {os = "win32"}
 ]
@@ -15,4 +16,5 @@
   [["ubuntu"] ["libblas-dev"]]
   [["centos"] ["blas-devel"]]
   [["fedora"] ["blas-devel"]]
+  [["freebsd"] ["blas" "gcc"]]
 ]
```

(I made it possible to override the compiler with `CC` variable to make it easier to debug the thing if something will break in the future).

Here's a similar change for `conf-lapack`, which has the same set of problems:

``` diff
--- ~/.opam/repo/default/packages/conf-lapack/conf-lapack.1/opam.orig  2016-08-31 16:13:36.000000000 +0300
+++ ~/.opam/repo/default/packages/conf-lapack/conf-lapack.1/opam       2016-08-31 16:13:40.000000000 +0300
@@ -6,7 +6,8 @@
 dev-repo: "https://github.com/ocaml/opam-repository.git"
 license: "GPL-2.0"
 build: [
-  ["sh" "-exc" "cc $CFLAGS test.c -llapack"] {os != "darwin" & os != "win32"}
+  ["sh" "-exc" "cc $CFLAGS test.c -llapack"] {os != "darwin" & os != "freebsd" & os != "win32"}
+  ["sh" "-exc" "${CC:-gcc} $CFLAGS test.c -llapack $LDFLAGS"] {os = "freebsd"}
   ["sh" "-exc" "cc -framework Accelerate $CFLAGS test.c -llapack"] {os = "darwin"}
   ["%{build}%/test-win.sh"] {os = "win32"}
 ]
@@ -15,4 +16,5 @@
   [["ubuntu"] ["liblapack-dev"]]
   [["centos"] ["lapack-devel"]]
   [["fedora"] ["lapack-devel"]]
+  [["freebsd"] ["lapack" "gcc"]]
 ]
```

With both of these changes I'm able to install `lacaml`, and it seems to work too.

(Sorry for providing patches inline without a pull request, but that is such a hassle).

### PR DESCRIPTION
**Pull request by mmottl**
Should fix #7330

### CODE REVIEW

File: packages/conf-blas/conf-blas.1/opam [hunk 1]
```diff
@@ -6,7 +6,8 @@ bug-reports: "https://github.com/ocaml/opam-repository/issues"
 dev-repo: "https://github.com/ocaml/opam-repository.git"
 license: "GPL-2.0"
 build: [
-  ["sh" "-exc" "cc $CFLAGS test.c -lblas"] {os != "darwin" & os != "win32"}
+  ["sh" "-exc" "cc $CFLAGS test.c -lblas"] {os != "darwin" & os != "freebsd" & os != "win32"}
+  ["sh" "-exc" "${CC:-gcc} $CFLAGS test.c -lblas $LDFLAGS"] {os = "freebsd"}
```
Comment (avsm) [inline packages/conf-blas/conf-blas.1/opam]: Hm, isn't that a `bash`-ism used there?  This logic might be better put into a shell script inside the package, as the same problem probably applies to several BSD variants (with /usr/local for -lblas)?

Comment (magv) [inline packages/conf-blas/conf-blas.1/opam]: That's pure [POSIX shell](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06_02).

GCC seems to include `/usr/local/{include,lib}` by default; at least under FreeBSD -- unlike system clang (but as explained in #7330, system clang is inappropriate here anyway).

**Comment by mmottl** [comment]
Sorry, I thought this had merged already, that's why the SQLite3 pull request ended up here, too.  I'll try to remember putting everything in separate branches next time.

**Comment by mmottl** [comment]
Not my best morning, forgot to run opam lint...

**Comment by yallop** [comment]
The OSX sqlite3 build is [failing](https://travis-ci.org/ocaml/opam-repository/jobs/163417030).

First there's this:

```
# The following new OS packages need to be installed: sqlite3
Warning: sqlite-3.14.2 already installed
# OS packages installation successful
```

and then this:

```
# Package sqlite3 was not found in the pkg-config search path.
# Perhaps you should add the directory containing `sqlite3.pc'
# to the PKG_CONFIG_PATH environment variable
# No package 'sqlite3' found
# E: Failure("Command ''/Users/travis/.opam/system/bin/ocamlbuild' lib/libsqlite3_stubs.a lib/dllsqlite3_stubs.so lib/sqlite3.cma lib/sqlite3.cmxa lib/sqlite3.a lib/sqlite3.cmxs -tag debug' terminated with error code 2")
```

**Comment by mmottl** [comment]
It seems this problem is related to `homebrew`.   @rleonid could you please check your `homebrew` configuration code in `myocamlbuild.ml`?  I don't use `homebrew` so I'm not quite sure what might be going on there.

**Comment by rleonid** [comment]
I've tried to replicate the problem on my `sqlite` fork and a travis build [here](https://travis-ci.org/rleonid/sqlite3-ocaml/jobs/165001542) and it seems like the "brew-checking" logic is working correctly. Maybe the error has to do with the 2nd (redundant?) installation of sqlite-3.14.2 ?

**Comment by mmottl** [comment]
@rleonid I just noticed in your Travis output that there are some issues with the SQLite distribution.  I have pushed two changes: one to get rid  of some warnings due to changes in the OCaml standard library, the other to make sure that proper finalization happens in a really old test case, which apparently now fails.  Your Travis failure is probably due to that failing test case.

**Comment by rleonid** [comment]
@mmottl What was the distribution issue that you noticed?

Yes, I noticed the tests were failing too. But that doesn't explain why the library doesn't build for `opam-repository` but does for my fork.

**Comment by mmottl** [comment]
@rleonid The build failure with the `opam-repository` appears unrelated, but it sure helps to fix other potential failures on the way.  I also don't understand why the `opam-repository` fails to build while your fork does.  Should I submit a new package with your new code to maybe get more info out of the `opam-repository` build, or is there anything simpler we can do?  Let me know if you first need to add even more diagnostics that would help us figure out what's going on.

**Comment by rleonid** [comment]
I only added a couple of `printf`'s to make sure that the output of the hack is sensible, it shouldn't be harmful. I can't think of any other check, the 2nd part of yallop's quote is `pkg-config` output, so we know that is present.  If you feel like going through the release-push-to-opam loop to try to figure out what's going on, then yes this seems like one way forward.

**Comment by mmottl** [comment]
The sqlite3 package seems to be passing the checks now.  Not sure what exactly made it work again.

**Comment by yallop** [comment]
Thanks for all your work on this!

**Comment by yallop** [comment]
I've rebased this onto `master` and squashed the sqlite commits (d572f7f).

**Pull request closed without merging**